monai/run_inference_single_image.py

```diff
@@ -0,0 +1,280 @@
+"""
+Script to run inference on a MONAI-based model for contrast-agnostic soft segmentation of the spinal cord.
+Prediction is stored in an independent folder given by subject name. The time taken for inference is stored in a json file.
+
+Usage:
+    python run_inference_single_image.py --path-img /path/to/my-awesome-SC-image.nii.gz --chkp-path /path/to/best/model 
+                --path-out /path/to/output/folder --crop-size <64x160x320> --device <cpu/gpu>
+
+Author: Naga Karthik
+
+"""
+
+import os
+import argparse
+import numpy as np
+from loguru import logger
+import torch.nn.functional as F
+import torch
+import json
+from time import time
+
+from models import create_nnunet_from_plans
+from monai.inferers import sliding_window_inference
+from monai.data import (DataLoader, CacheDataset, load_decathlon_datalist, decollate_batch)
+from monai.transforms import (Compose, EnsureTyped, Invertd, SaveImage, Spacingd,
+                              LoadImaged, NormalizeIntensityd, EnsureChannelFirstd, 
+                              DivisiblePadd, Orientationd, ResizeWithPadOrCropd)
+
+
+# NNUNET global params
+INIT_FILTERS=32
+ENABLE_DS = True
+
+nnunet_plans = {
+    "UNet_class_name": "PlainConvUNet",
+    "UNet_base_num_features": INIT_FILTERS,
+    "n_conv_per_stage_encoder": [2, 2, 2, 2, 2, 2],
+    "n_conv_per_stage_decoder": [2, 2, 2, 2, 2],
+    "pool_op_kernel_sizes": [
+        [1, 1, 1],
+        [2, 2, 2],
+        [2, 2, 2],
+        [2, 2, 2],
+        [2, 2, 2],
+        [1, 2, 2]
+    ],
+    "conv_kernel_sizes": [
+        [3, 3, 3],
+        [3, 3, 3],
+        [3, 3, 3],
+        [3, 3, 3],
+        [3, 3, 3],
+        [3, 3, 3]
+    ],
+    "unet_max_num_features": 320,
+}
+
+
+def get_parser():
+
+    parser = argparse.ArgumentParser(description="Run inference on a MONAI-trained model")
+
+    parser.add_argument("--path-img", type=str, required=True,
+                        help="Path to the image to run inference on")
+    parser.add_argument("--chkp-path", type=str, required=True, help="Path to the checkpoint folder")
+    parser.add_argument("--path-out", type=str, required=True, 
+                        help="Path to the output folder where to store the predictions and associated metrics")
+    parser.add_argument('-crop', '--crop-size', type=str, default="64x160x320", 
+                        help='Patch size used for center cropping the images during inference. Values correspond to R-L, A-P, I-S axes'
```

Comment:
```suggestion
                        help='Size of the window used to crop the volume before inference (and after resampling to 1mm isotropic). The window is centered in the middle of the volume. Dimensions are in the order R-L, A-P, I-S.'
```

Reply:
I find it quite difficult from the user standpoint to know how to crop, if the dimensions and resolution are different from their native image. Could we consider maybe cropping before resampling and reorientation? To discuss…

Reply:
good for now-- pls open issue so we can discuss this once we have more projects from the lab where deepseg is being used-- and crossref this comment

Reply:
Done here #70